zallet/src/components/sync.rs

```diff
@@ -329,3 +345,220 @@ async fn recover_history(
+    interval.tick().await;
+
+    // TODO: Remove this once https://github.com/zcash/librustzcash/issues/1805
+    // is implemented and used by Zallet.
+    let mut hotloop_preventer = time::interval(Duration::from_secs(1));
+    loop {
+        hotloop_preventer.tick().await;
+
+        let requests = db_data.transaction_data_requests()?;
+        if requests.is_empty() {
+            // Wait for new requests.
+            debug!("No transaction data requests, sleeping");
+            interval.tick().await;
+            continue;
+        }
+
+        for request in requests {
+            match request {
+                TransactionDataRequest::GetStatus(txid) => {
+                    info!("Getting status of {txid}");
+                    let status = match chain.get_raw_transaction(txid.to_string(), Some(1)).await {
+                        // TODO: Zaino should have a Rust API for fetching tx details,
+                        // instead of requiring us to specify a verbosity and then deal
+                        // with an enum variant that should never occur.
+                        Ok(zebra_rpc::methods::GetRawTransaction::Raw(_)) => unreachable!(),
+                        Ok(zebra_rpc::methods::GetRawTransaction::Object(tx)) => tx
+                            .height
+                            .map(BlockHeight::from_u32)
+                            .map(TransactionStatus::Mined)
+                            .unwrap_or(TransactionStatus::NotInMainChain),
+                        // TODO: Detect error corresponding to
+                        // `TransactionStatus::TxidNotRecognized`.
+                        Err(e) => return Err(e.into()),
+                    };
+
+                    db_data.set_transaction_status(txid, status)?;
+                }
+                TransactionDataRequest::Enhancement(txid) => {
+                    info!("Enhancing {txid}");
+                    let tx = match chain.get_raw_transaction(txid.to_string(), Some(1)).await {
+                        // TODO: Zaino should have a Rust API for fetching tx details,
+                        // instead of requiring us to specify a verbosity and then deal
+                        // with an enum variant that should never occur.
+                        Ok(zebra_rpc::methods::GetRawTransaction::Raw(_)) => unreachable!(),
+                        Ok(zebra_rpc::methods::GetRawTransaction::Object(tx)) => {
+                            let mined_height = tx.height.map(BlockHeight::from_u32);
+
+                            // TODO: Zaino should either be doing the tx parsing for us,
+                            // or telling us the consensus branch ID for which the tx is
+                            // treated as valid.
+                            // TODO: We should make the consensus branch ID optional in
+                            // the parser, so Zaino only need to provide it to us for v4
+                            // or earlier txs.
```

Comment:
This is actually the first circumstance where it matters; everywhere else, we've opted to just use the "wrong" consensus branch ID because we effectively ignore it for any operation on v4 transactions. If we make it optional for the parser, we just move the problem; then, `TransactionData` makes it possible to represent the invalid state of a v5 transaction that has no consensus branch ID set.